components/fxa-client/src/internal/state_manager.rs

```diff
@@ -157,23 +169,59 @@ impl StateManager {
     ) {
         // When our keys change, we might need to re-register device capabilities with the server.
         // Ensure that this happens on the next call to ensure_capabilities.
-        self.persisted_state.device_capabilities.clear();
+        self.clear_server_local_device_info();
 
         for (scope, key) in scoped_keys {
             self.persisted_state.scoped_keys.insert(scope, key);
         }
         self.persisted_state.refresh_token = Some(refresh_token);
         self.persisted_state.session_token = session_token;
+        self.persisted_state.logged_out_from_auth_issues = false;
         self.flow_store.clear();
     }
 
     /// Called when the account is disconnected.  This clears most of the auth state, but keeps
     /// some information in order to eventually reconnect to the same user account later.
     pub fn disconnect(&mut self) {
-        self.persisted_state = self.persisted_state.start_over();
+        self.persisted_state.current_device_id = None;
+        self.persisted_state.refresh_token = None;
+        self.persisted_state.scoped_keys = HashMap::new();
+        self.persisted_state.last_handled_command = None;
+        self.persisted_state.commands_data = HashMap::new();
+        self.persisted_state.access_token_cache = HashMap::new();
+        self.persisted_state.device_capabilities = HashSet::new();
+        self.persisted_state.server_local_device_info = None;
+        self.persisted_state.session_token = None;
+        self.persisted_state.logged_out_from_auth_issues = false;
+        self.flow_store.clear();
+    }
+
+    /// Called when the account is logged out because of auth issues.  This clears the auth state,
+    /// but leaves `current_device_id` untouched.  That way, if the user re-authenticates they can
+    /// continue to use their existing device entry.
+    pub fn logout_from_auth_issues(&mut self) {
```

Comment:
I don't like this name tbh - I don't see it as the user being logged out, but instead being in a bad state. For example, do we really want to reset `last_handled_command` here? If a user on desktop changes their password then later sends a tab to this device (forgetting they are yet to re-auth it), I think we'd want this device to see the tab immediately after reauthenticating.  Something like "update_for_auth_issues()" or similar seems better - but I don't want to block landing on this if it ends up being a PITA.

(I guess you could argue that this is the only method called "log out" and can't be confused with "disconnect()", which I also can't argue with :)

Also: it's not completely clear how the app responds once the user reauths. Can we be *sure* it will re-call `set_device_capabilities()` - given that is about the *device* state and independent of the *user* state, I can see that it might not, so maybe this should not be cleared too? Ditto the device name - Tarik mentions the device ID should remain unchanged, can we be sure the app will do whatever it does to ensure the device name also doesn't change?

Reply:
I don't really like the name either, it's hard to pick one.  How about `transition_to_auth_issues()`?  I don't love that one either, but it makes it clear that it's a state-transition at least.

Good point on `last_handled_command`.  I'll make it so it doesn't reset that.  I'm also thinking `device_capabilities` shouldn't be reset, especially since this PR changes the semantics on that a bit.  It's now only used to know store what the consumer app last sent to `ensure_capabilities` or `initialize_device`.  Seems like that should persist through auth issues.

Reply:
`note_auth_issue()` maybe?

Reply:
or "observe", or "on"?

Reply:
Thanks for the naming help.  I went with `on_auth_issues`.

The method currently resets the same fields as `disconnect`, except `current_device_id`, `device_capabilities`, and `last_handled_command`.  Does that seem right?

Reply:
That seemed right to me when I was looking over them.